h2o-py/h2o/fusion.py

```diff
@@ -0,0 +1,122 @@
+# -*- encoding: utf-8 -*-
+"""
+Rapids expression fusion.
+
+:copyright: (c) 2016 H2O.ai
+:license:   Apache License Version 2.0 (see LICENSE for details)
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import h2o.expr
+
+class Fusion(object):
+    #
+    # Generic fusion representation
+    #
+    def __init__(self, supported_ops):
+        self._supported_ops = supported_ops
+
+    def supports(self, op):
+        # Supports given operator
+        return op in self._supported_ops
+
+    def is_applicable(self, expr):
+        # Is fusion applicable for given operator
+        return False
+
+    def get_fusion(self, expr):
+        # Return fusion which is mapping from ExprNode to ExprNode
+        return id(expr)
+
+class FoldFusion(Fusion):
+    """
+    Fold fusion: support operators which
+    accepts array of parameters (e.g., append, cbind):
+
+    For example: append dst (src col_name)+
+      (append (append dst srcX col_name_Y) src_A col_name_B) is transformed to
+      (append dst src_X col_name_Y src_A col_name_B)
+
+    Objective:
+      - the folding save a temporary variable during evaluation
+    """
+    def __init__(self):
+        super(self.__class__, self).__init__(["append", "cbind"])
+
+    def is_applicable(self, expr):
+        # Only applicable if the source parameter is the same operator
+        assert isinstance(expr, h2o.expr.ExprNode)
+        assert any(expr._children)
+        return expr._children[0]._op == expr._op
+
+    def get_fusion(self, expr):
+        def fusion_fce(ctx):
+            nested_expr = expr.arg(0)
+            expr._children = nested_expr._children + expr._children[1:]
+            return expr
+        return fusion_fce
+
+class SkipFusion(Fusion):
```

Comment:
It is probably possible that the rapid expression will modify the frame you are working with (eg. AstAssign), in that case you could be using incorrect information to optimize the expression.

Reply:
Yup, skip fusion is design under assumption that frame cache is properly maintained.

Reply:
It is not just the frame cache. The side effect can happen in the rapid expression itself.

Reply:
Eg. it is possible that someone will use AstAssign to overwrite the frame "frame_with_100_columns" in

```
(append (assign frame_with_100_columns completely_different frame) (col_py (append frame_with_100_columns dummy_col dummy_name) 1))
```

Your optimizer would simplify the sub-expression using wrong assumption on the frame.

Reply:
My suggestion would be to implement this kind of optimizer on the backend. Some optimizations can be done on client side to make the expression simpler. But we can still optimize on the backend. This would allow us to have the optimizations both for R/Python and any other client.

Reply:
regarding assign  - good point, we are not invalidating references 

regarding opt at backend - agree

Reply:
We can disable SKipFusion for now, or just use it for named frames - not perfect solution, but at least it helps a lot for use-cases a la pubdev-4873